### DISCUSSION TITLE
BoundsError access Bool

### DISCUSSION THREAD

**Discussion by AnHeuermann** [ThummeTo/FMI.jl#65] [General]
## Description

When I simulate a FMU generated from the Modelica Standard Library I get a bounds error.

## How to reproduce

Consider the following minimal example with a CS-ME 2.0 FMU `Modelica.Mechanics.Translational.Examples.Damper.fmu` generatet from OpenModelica (OMCompiler v1.19.0-dev.505+ga499198eb3):

```julia
using FMI

myFMU = fmiLoad("Modelica.Mechanics.Translational.Examples.Damper.fmu")
fmiInstantiate!(myFMU)
success, simData = fmiSimulate(myFMU, 0.0, 1.0)
```

The error message:
```julia
julia> success, simData = fmiSimulate(myFMU, 0.0, 1.0)
ERROR: BoundsError: attempt to access Bool at index [2]
Stacktrace:
 [1] indexed_iterate(I::Bool, i::Int64, state::Nothing)
   @ Base ./tuple.jl:98
 [2] top-level scope
   @ REPL[4]:1
```

OMSimulator can simulate the FMU fine, so I guess it is not an issue with the FMU. But I didn't test a non-OpenModelica tool on it.

## Additional files

Here is the FMU (remove the .zip file extension) with Linux binaries: 
[Modelica.Mechanics.Translational.Examples.Damper.fmu.zip](https://github.com/ThummeTo/FMI.jl/files/7837780/Modelica.Mechanics.Translational.Examples.Damper.fmu.zip)

--------------------------

## Version 'n stuff
OS: Ubuntu 20.04
Julia:
```
Julia Version 1.7.1
Commit ac5cc99908 (2021-12-22 19:35 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-10700KF CPU @ 3.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
```
```
(@v1.7) pkg> status FMI
      Status `~/.julia/environments/v1.7/Project.toml`
  [14a09403] FMI v0.3.6
```
OpenModelica: OMCompiler v1.18.1 and v1.19.0-dev.505+ga499198eb3

**Comment by AnHeuermann** [comment]
Okay, I just called the `simulate` function wrong. From the README I got that I can call it like
```julia
success, simData = fmiSimulate(myFMU, 0.0, 10.0; recordValues=["mass.s"])
```
but when I remove `recordValues` from the call the return value will change and only return one bool `success`. And that is the reason for the error.

Is this a good idea to have the function interface change on a optional argument? I expected it to save all variables when I don't specifically tell fmiSimulate what I want to save, but instead it saved nothing.

**Comment by ThummeTo** [comment]
Believe it or not: We already had this discussion internally :-D

Should a function always return the same pattern? There are pros and cons, but we decided to don't return "dummy"-values if they are not needed/used. On the other hand many languages don't support this return feature, so for developers from these languages this behaviour is a bit unexpected. We decided to use the language features of Julia in this case. (you can call `?fmi2Simulate`, the keyword arguments and return values are explained there)

But I am open for other opinions of course!

BTW: I will move this to discussions.

**Reply by AnHeuermann** [reply]
I looked into the help for `fmiSimulate`, but it wasn't returning the information I needed.

```julia
help?> fmiSimulate
search: fmiSimulate fmiSimulateME fmiSimulateCS fmi2Simulate fmi2SimulateME fmi2SimulateCS fmi2SimulationResult fmi2SimulationResultGetTime fmi2SimulationResultGetValues fmi2SimulationResultGetValuesAtIndex fmiIsCoSimulation

  Simulate an fmu according to its standard from 0.0 to t_stop.
```

**Reply by AnHeuermann** [reply]
I think the doc string could be improved. For a lot of Julia base functions I get the function call returned first. e.g. for `?sort` I get:
```
sort(v; alg::Algorithm=defalg(v), lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward)
```

but the relevant information for fmi2Simulate is harder to find:
```
help?> fmi2Simulate
search: fmi2Simulate fmi2SimulateME fmi2SimulateCS fmi2SimulationResult fmi2SimulationResultGetTime fmi2SimulationResultGetValues fmi2SimulationResultGetValuesAtIndex fmi2IsCoSimulation

  Starts a simulation of the fmu instance for the matching fmu type. If both types are available, CS is preferred over ME.

  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

  Starts a simulation of the FMU instance for the matching FMU type, if both types are available, CS is preferred.

  Keywords: - recordValues: Array of variables (strings or variableIdentifiers) to record. Results are returned as DiffEqCallbacks.SavedValues - setup: Boolean, if FMU should be setup (default=true) - reset: Boolean, if FMU should
  be reset before simulation (default reset=setup) - inputValues: Array of input variables (strings or variableIdentifiers) to set at every simulation step - inputFunction: Function to retrieve the values to set the inputs to -
  saveat: [ME only] Time points to save values at (interpolated) - solver: [ME only] Any Julia-supported ODE-solver (default is Tsit5) - customFx: [ME only, deperecated] Ability to give a custom state derivative function ẋ=f(x,t)

  Returns: - success::Bool for CS-FMUs - ODESolution for ME-FMUs - if keyword recordValues is set, a tuple of type (success::Bool, DiffEqCallbacks.SavedValues) for CS-FMUs - if keyword recordValues is set, a tuple of type
  (ODESolution, DiffEqCallbacks.SavedValues) for ME-FMUs
```

**Reply by ThummeTo** [reply]
Yes, you are right, they could be much better, I will open an issue for this: #67
